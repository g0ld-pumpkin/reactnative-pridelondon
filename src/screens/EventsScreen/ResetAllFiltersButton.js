// @flow
import React from "react";
import { Animated, StyleSheet, Easing, InteractionManager } from "react-native";
import type { ViewLayoutEvent } from "react-native/Libraries/Components/View/ViewPropTypes";
import FilterHeaderButton from "./FilterHeaderButton";
import text from "../../constants/text";

type Props = {
  visible: boolean,
  onPress: () => void,
  animationTime: number,
  animationDelay: number
};

type State = {
  isAnimating: boolean
};

const DEFAULT_HEIGHT = 120;
const DEFAULT_FADE_VALUE = 1;
const DEFAULT_TOP_OFFSET_VALUE = 0;

class ResetAllFiltersButton extends React.PureComponent<Props, State> {
  static defaultProps = {
    animationTime: 300,
    animationDelay: 0
  };

  constructor(props: Props) {
    super(props);
    this.state = {
      isAnimating: false
    };
  }

  componentWillUnmount() {
    this.animationFinished();
  }

  setButtonHeight = (e: ViewLayoutEvent): void => {
    const { height } = e.nativeEvent.layout;
    this.height = height;
  };

  fadeOut = (): void => {
    const { animationTime, animationDelay } = this.props;

    Animated.parallel([
      Animated.timing(this.fadeValue, {
        toValue: 0,
        duration: animationTime / 2,
        delay: animationDelay
      }),
      Animated.timing(this.topOffset, {
        toValue: -this.height,
        duration: animationTime,
        easing: Easing.out(Easing.quad),
        delay: animationDelay
      })
    ]).start(this.animationFinished);
  };

  resetAllFilters = (): void => {
    this.props.onPress();
    InteractionManager.runAfterInteractions(() => {
      this.setState({ isAnimating: true });
      this.fadeOut();
    });
  };

  resetAnimation = (): void => {
    this.topOffset = new Animated.Value(DEFAULT_TOP_OFFSET_VALUE);
    this.fadeValue = new Animated.Value(DEFAULT_FADE_VALUE);
  };

  animationFinished = (): void => {
    if (this.state.isAnimating) {
      this.setState({ isAnimating: false });
      this.resetAnimation();
    }
  };

  topOffset: Animated.Value = new Animated.Value(DEFAULT_TOP_OFFSET_VALUE);
  fadeValue: Animated.Value = new Animated.Value(DEFAULT_FADE_VALUE);
  height: number = DEFAULT_HEIGHT;

  render() {
    const isVisible: boolean = this.state.isAnimating || this.props.visible;
    return (
      isVisible && (
        <Animated.View
          onLayout={this.setButtonHeight}
          style={[
            styles.clearAllWrapper,
            { opacity: this.fadeValue, marginTop: this.topOffset }
          ]}
        >
          <FilterHeaderButton
            active={false}
            text={text.resetAllFilters}
            label={text.resetAllFilters}
            style={styles.clearAll}
            onPress={this.resetAllFilters}
          />
        </Animated.View>
      )
    );
  }
}

const styles = StyleSheet.create({
  clearAll: {
    minHeight: 0,
    paddingTop: 16
  },
  clearAllWrapper: {
    flexDirection: "row",
    justifyContent: "flex-end"
  }
});

export default ResetAllFiltersButton;
