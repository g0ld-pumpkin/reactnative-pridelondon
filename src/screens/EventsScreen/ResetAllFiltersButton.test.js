// @flow
import React from "react";
import { Animated, InteractionManager } from "react-native";
import { shallow } from "enzyme";
import ResetAllFiltersButton from "./ResetAllFiltersButton";
import FilterHeaderButton from "./FilterHeaderButton";

const noOp = () => {};

let runAfterInteractionsSpy;
let animatedParallelSpy;
let animatedTimingSpy;

beforeEach(() => {
  runAfterInteractionsSpy = jest
    .spyOn(InteractionManager, "runAfterInteractions")
    .mockImplementation(noOp);
  animatedParallelSpy = jest
    .spyOn(Animated, "parallel")
    .mockImplementation(noOp);
  animatedTimingSpy = jest.spyOn(Animated, "timing").mockImplementation(noOp);
});

afterEach(() => {
  runAfterInteractionsSpy.mockRestore();
  animatedParallelSpy.mockRestore();
  animatedTimingSpy.mockRestore();
});

it("renders correctly", () => {
  const output = shallow(<ResetAllFiltersButton visible onPress={noOp} />);
  expect(output).toMatchSnapshot();
});

it("does not render when visible=false", () => {
  const output = shallow(
    <ResetAllFiltersButton visible={false} onPress={noOp} />
  );
  expect(output).toMatchSnapshot();
});

it("calls onPress method when pressed", () => {
  const mockOnPress = jest.fn();
  const output = shallow(
    <ResetAllFiltersButton
      visible
      animationTime={0}
      animationDelay={0}
      onPress={mockOnPress}
    />
  );

  output.find(FilterHeaderButton).simulate("press");

  expect(runAfterInteractionsSpy).toHaveBeenCalled();
  expect(mockOnPress).toHaveBeenCalled();
});

it("starts animation after being pressed", () => {
  const output = shallow(
    <ResetAllFiltersButton
      visible
      onPress={noOp}
      animationTime={0}
      animationDelay={0}
    />
  );

  const animation = { start: jest.fn() };
  animatedParallelSpy.mockReturnValue(animation);

  output.find(FilterHeaderButton).simulate("press");

  // Animation should start after interaction.
  const interactionCallback = runAfterInteractionsSpy.mock.calls[0][0];
  interactionCallback();

  expect(output.state().isAnimating).toBe(true);
  expect(animatedParallelSpy).toHaveBeenCalled();
  expect(animatedTimingSpy).toHaveBeenCalledTimes(2);
  expect(animation.start).toHaveBeenCalled();

  // Callback is called when animation is completed.
  const completedCallback = animation.start.mock.calls[0][0];
  completedCallback();

  expect(output.state().isAnimating).toBe(false);
  expect(output.instance().fadeValue).toEqual(new Animated.Value(1));
  expect(output.instance().topOffset).toEqual(new Animated.Value(0));
});

it("resets animation on unmount", () => {
  const output = shallow(
    <ResetAllFiltersButton
      visible
      onPress={noOp}
      animationTime={0}
      animationDelay={0}
    />
  );

  const animation = { start: jest.fn() };
  animatedParallelSpy.mockReturnValue(animation);

  output.find(FilterHeaderButton).simulate("press");

  // Animation should start after interaction.
  const interactionCallback = runAfterInteractionsSpy.mock.calls[0][0];
  interactionCallback();

  expect(output.state().isAnimating).toBe(true);

  // Unmount.
  output.unmount();

  // expect(output.state().isAnimating).toBe(false);
  // expect(output.instance().fadeValue).toEqual(new Animated.Value(1));
  // expect(output.instance().topOffset).toEqual(new Animated.Value(0));

  // Completion callback does not make it crash.
  const completedCallback = animation.start.mock.calls[0][0];
  completedCallback();
});

it("sets header height for use with animation", () => {
  const output = shallow(<ResetAllFiltersButton visible onPress={noOp} />);
  const mockEvent = {
    nativeEvent: { layout: { height: 100 } }
  };
  output.instance().setButtonHeight(mockEvent);

  expect(output.instance().height).toEqual(100);
});
