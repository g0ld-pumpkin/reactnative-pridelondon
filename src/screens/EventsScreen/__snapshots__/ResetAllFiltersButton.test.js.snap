// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`does not render when visible=false 1`] = `""`;

exports[`renders correctly 1`] = `
<AnimatedComponent
  onLayout={[Function]}
  style={
    Array [
      Object {
        "flexDirection": "row",
        "justifyContent": "flex-end",
      },
      Object {
        "marginTop": 0,
        "opacity": 1,
      },
    ]
  }
>
  <FilterHeaderButton
    active={false}
    badgeValue={null}
    label="Reset all filters"
    onPress={[Function]}
    style={
      Object {
        "minHeight": 0,
        "paddingTop": 16,
      }
    }
    text="Reset all filters"
  />
</AnimatedComponent>
`;
